### DISCUSSION TITLE
AI Thinker v2.2 (A202) Headphones volume controls

### DISCUSSION THREAD

**Discussion by billy8119** [pschatzmann/arduino-audiokit#26] [General]
The headphone volume is supposed to be adjustable, yeah?
I poked the volume in the main arduino sketch, (set it to 90), and also enabled AI_THINKER_ES8388_VOLUME_HACK, but nothing.
Meanwhile, I just want to ensure I know what's going on in the code (sorry for nub question) but the code in the hack:
res = es_write_reg(ES8388_ADDR, ES8388_DACCONTROL4, volume);
  res |= es_write_reg(ES8388_ADDR, ES8388_DACCONTROL5, volume);
Is trying to set Register 4 of the ES8388, known as "DAC Power Management", yeah? If so, this is a bit-field, and setting it to just any old volume integer from 0-100 is likely to cause unexpected behavior, yeah? Or is DACCONTROL4 NOT, in fact, "Register 4" as documented in the Everest 8388 DS.pdf?

**Comment by pschatzmann** [comment]
I assume that you were already setting the volume in the AudioKitStream to the maximum: kit.setVolume(100); 
Please note the following defines:
#define ES8388_DACCONTROL4      0x1a
#define ES8388_DACCONTROL5      0x1b
So this is Register 26 and 27 which is in the [Datasheet](https://datasheet.lcsc.com/lcsc/1912111437_Everest-semi-Everest-Semiconductor-ES8388_C365736.pdf) DAC Control 4 and 5. I see the chance to scale the value here a bit since the maximum is binary 11000000 (=192) but the API is limited to 100

Are your speakers 4Ohm  ?

**Reply by billy8119** [reply]
To my profound shame, I realized that I should RTFM and answer my own question, yes.  The mapping of labels to actual registers is right there in plain English in arduino-audiokit-main/src/audio_driver/es8388/es8388.h  Oops. Sorry.
Meanwhile, am I crazy, or is "0" supposed to be the loudest, and 192 the quietest?  Docs suggest "Digital volume control attenuates the signal in 0.5 dB incremental from 0 to –96 dB." ... it's not BOOSTING the volume as the input goes up, it's ATTENUATING it.  I mean, I still have no effect on volume whatsoever, regardless of what values I try... Thanks for the response though.

**Comment by pschatzmann** [comment]
I agree that the volume hack will actually make the volume worse. The only explanation that I would have that there is maybe a bug in some of the es8388 where this is messed up. What is your number next to the V2.2 ?

Which sketch are you using to test ?

**Reply by billy8119** [reply]
It's silk-screened A202.  I see you have 2444.  I can't find any documentation enumerating the various sub-versions and their differences.  Am I missing the obvious?
I'm using https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sd-audiokit/player-sd-audiokit.ino

**Comment by pschatzmann** [comment]
Please note that both the Kit and Player have some independent volume control. So set
``` 
kit.setVolume(100);
player.setVolume(1.0);
```
for max volume...

I was just testing with the AI_THINKER_ES8388_VOLUME_HACK 1 on my board and the volume is slightly attenuated. So this should not really be used on the 3478 and I guess A202 boards as well...

**Reply by billy8119** [reply]
Interestingly, the addition of kit.setVolume(100);  ..the signal is significantly WEAKER.  I also tried kit.setVolume(1), and there was no change - still a very weak signal.  (Barely audible on headphones) I'll go try stuff.  
(At least there's a volume change, which is the first time that's happened, so thanks!)

**Comment by pschatzmann** [comment]
Strange on my board this setting is really loud...
Did you deactivate the volume hack ?

**Reply by billy8119** [reply]
Things I've tried:
A) Leave volume hack, and just insert kit.setVolume  -=> Volume on headphones is now very quiet
B) use kit.setVolume(1); -=> No change, still quiet
C) remove kit.setVolume altogether -=> No change .. still quiet (!?!?)
D) Remove volume hack, and use code exactly as it was from the "Factory" -=> No change... still quiet... called doctor for a hearing test...
E) Doctor says hearing is fine.  Tells me to stop being a hypochondriac

**Reply by billy8119** [reply]
(also tried resetting the card with hard power off... just in case it's a random initialization thing... no change, still ultra quiet.

**Comment by pschatzmann** [comment]
Oh, I tested only the speaker volume - not the headphones...

**Reply by billy8119** [reply]
I'm not 100% certain the kit opamps are hooked up to line out (headphones).. but I should still be able to adjust volume with the esp8388.  I'll keep tryin' stuff.

**Comment by pschatzmann** [comment]
Just tested with my earphones from my mobile phone and with max volume it's also quite loud. 
Would you know if earphones and your headphones have the same impedence (ohm values) ?
Did you try with earphones as well ?

**Reply by billy8119** [reply]
Well, I don't have earphones for my mobile phone, but I DID plug these particular headphones into my phone, and yes, it can be quite loud too.  I assume it's a normal 8ohm thing.  Maybe 4ohm.  I bet it's normal. I also plugged this kit board into a little AMP, and it's *barely* audible. So, prolly not just the headphones.

**Comment by pschatzmann** [comment]
Strange. I tested with some headphones on my board and the volume at max is good.
I start to get the impression that there is something wrong with your board. Maybe it is damaged. 

It would be interresting to hear if others with a A202 have the same problem...

**Reply by billy8119** [reply]
Out of curiosity, KIT_LOGD("es8388_set_voice_volume: %d", volume);
should print "es8388_set_voice_volume: xxx" at some point in the USB "console" stream, right?
IE that's what the KIT_LOGD macro does?
Well, I never see that output.  I only see [I] AudioStreams.h : 799 - setVolume: 0.700000
when songs change.  Does this mean the volume is actually not being set in this code? The hack might have never ever run, then, as I never see this log message...
And yes, I am not ruling out damage to my board.  Still, damage or no damage, I should have some CONTROL over volume, yeah? So far, nothing.  Just the one blip where everything grew more quiet after the first kit.setVolume recompile...

**Comment by pschatzmann** [comment]
Yes, If you set the log level for the kit to debug. By default it is set to wrning only...
I suggest to use a [simpler sketch for testing](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-url_mp3-audiokit/streams-url_mp3-audiokit.ino) so that you only need to bother about the volume control of the kit.

**Comment by billy8119** [comment]
I laugh at how fabulous your code is.  LOL. One line for "URLStream" and it's all set. LOL Well done.  And yes, thanks, this shaves off 50% on compile time.

**Comment by billy8119** [comment]
New sketch: the only object that has setVolume is i2s, and calling setVolume on i2s doesn't result in any invocation of es8388_set_voice_volume()... how SHOULD I be setting volume in this sketch? Thanks.

**Comment by pschatzmann** [comment]
That can't be.  I just reviewed the logic AudioKitStream.setVolume is calling AudioKit.setVolume() which is calling audio_hal_set_volume() which is linked to es8388_set_voice_volume()

Did you set the log level with
LOGLEVEL_AUDIOKIT = AudioKitDebug; 

Did you also define your board in AudioKitSettings.h ?

**Comment by billy8119** [comment]
I swapped the code to KIT_LOGI ..same thing?

On June 25, 2022 4:58:58 p.m. EDT, Phil Schatzmann ***@***.***> wrote:
>That can't be.  I just reviewed the logic and I am ending calling es8388_set_voice_volume()
>Did you set the log level with
>LOGLEVEL_AUDIOKIT = AudioKitDebug; 
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/pschatzmann/arduino-audiokit/discussions/26#discussioncomment-3023987
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>
-- 
Sent from mobile

**Reply by billy8119** [reply]
I was using KIT_LOGI instead of LOGD, I thought it would work, but it doesn't.  Switched to adding the AudioKitDebug as you suggest, and that works fine.
More research has suggested some oddities: on my card, when setting setVolume(0.5) the "volume" value that gets programmed into reg46 and reg47 was only "6".  That equates to -39dB, which is 6dB louder than the quietest the es8388 can actually do.  I hard-inserted the max volume (33, which is +4.5dB) and this works OK.  No distortion.  Still fairly quiet on a normal amp when using line out, but yes, perhaps that's a combination of my actual mp3 file, and the board itself.  I have not yet tried a sine wave using -32767 <-> +32768 at 16 bits or anything.
What input file were you using that was really loud? Is it legal to share this .mp3 clip? I'd kinda like to know if part of the problem is my chosen file(s).
Then I'll have a closer look at volume values and see if there's a bug in there somewhere, or if that's just how it's supposed to work.

**Comment by pschatzmann** [comment]
Wait: the volume scale for the AudioKit is from 1 to 100: So max volume is 100. This logic comes from the original Espressif code but I plan to add support for float values as well to be consistent with my AudioKit volumes where 1.0 is max.

I am using the streams-url-audiokit.ino sketch and the speaker volume is rather load. With headphones it is just slightly higher then what I would want to use.

With 100 I am getting 
```
05:58:19.630 -> rst:0x1 (POWERON_RESET),boot:0x1f (SPI_FAST_FLASH_BOOT)
05:58:19.630 -> configsip: 0, SPIWP:0xee
05:58:19.630 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
05:58:19.630 -> mode:DIO, clock div:1
05:58:19.630 -> load:0x3fff0030,len:1344
05:58:19.668 -> load:0x40078000,len:13516
05:58:19.668 -> load:0x40080400,len:3604
05:58:19.668 -> entry 0x400805f0
05:58:20.048 -> Debug:   audio_hal_set_volume: 100
05:58:20.048 -> Error:   audio_hal handle is null
05:58:20.048 -> [I] AudioTypes.h : 48 - sample_rate: 44100
05:58:20.048 -> [I] AudioTypes.h : 49 - channels: 2
05:58:20.081 -> [I] AudioTypes.h : 50 - bits_per_sample: 16
05:58:20.081 -> [I] I2SConfig.h : 95 - rx/tx mode: TX
05:58:20.081 -> [I] I2SConfig.h : 96 - port_no: 0
05:58:20.081 -> [I] I2SConfig.h : 97 - is_master: Master
05:58:20.081 -> [I] I2SConfig.h : 98 - sample rate: 44100
05:58:20.081 -> [I] I2SConfig.h : 99 - bits per sample: 16
05:58:20.081 -> [I] I2SConfig.h : 100 - number of channels: 2
05:58:20.081 -> [I] I2SConfig.h : 101 - i2s_format: I2S_STD_FORMAT
05:58:20.081 -> [I] I2SConfig.h : 103 - auto_clear:0
05:58:20.081 -> [I] I2SConfig.h : 110 - buffer_count:10
05:58:20.117 -> [I] I2SConfig.h : 111 - buffer_size:512
05:58:20.117 -> [I] I2SConfig.h : 115 - pin_bck: 14
05:58:20.117 -> [I] I2SConfig.h : 116 - pin_ws: 15
05:58:20.117 -> [I] I2SConfig.h : 117 - pin_data: -1515870811
05:58:20.117 -> [W] AudioKit.h : 46 - sd_active = true
05:58:20.117 -> Info:    bool AudioKit::begin(AudioKitConfig)
05:58:20.117 -> Info:    Selected board: 5
05:58:20.117 -> Info:    headphone detection is active
05:58:20.117 -> Debug:   audio_hal_init
05:58:20.117 -> Debug:   audio_calloc: 1 * 36
05:58:20.155 -> Debug:   es8388_init
05:58:20.155 -> Debug:   i2c_init
05:58:20.155 -> Info:    i2c port: 0
05:58:20.155 -> Info:    i2c sda: 33
05:58:20.155 -> Info:    i2c scl: 32
05:58:20.155 -> Debug:   void* i2c_bus_create(i2c_port_t, i2c_config_t*)
05:58:20.155 -> Info:    i2c sda: 33
05:58:20.155 -> Info:    i2c scl: 32
05:58:20.155 -> Info:    i2c clk_speed: 100000
05:58:20.155 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=25 value=4
05:58:20.155 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=1 value=80
05:58:20.155 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=2 value=0
05:58:20.188 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=53 value=160
05:58:20.188 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=55 value=208
05:58:20.188 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=57 value=208
05:58:20.188 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=8 value=0
05:58:20.188 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=4 value=192
05:58:20.226 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=0 value=18
05:58:20.226 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=23 value=24
05:58:20.226 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=24 value=2
05:58:20.226 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=38 value=0
05:58:20.226 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=39 value=144
05:58:20.226 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=42 value=144
05:58:20.260 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=43 value=128
05:58:20.260 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=45 value=0
05:58:20.260 -> Debug:   es8388_set_adc_dac_volume: 0
05:58:20.260 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=27 value=0
05:58:20.260 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=26 value=0
05:58:20.260 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=4 value=60
05:58:20.298 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=3 value=255
05:58:20.298 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=9 value=187
05:58:20.298 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=10 value=0
05:58:20.298 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=11 value=2
05:58:20.298 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=12 value=13
05:58:20.331 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=13 value=2
05:58:20.331 -> Debug:   es8388_set_adc_dac_volume: 0
05:58:20.331 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=16 value=0
05:58:20.331 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=17 value=0
05:58:20.331 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=3 value=9
05:58:20.331 -> Debug:   es8388_pa_power: true
05:58:20.365 -> Debug:   audio_codec_initialize -> 0
05:58:20.365 -> Info:    audio_codec_initialize-END-OK
05:58:20.365 -> Debug:   es8388_config_i2s
05:58:20.365 -> Debug:   es8388_config_fmt
05:58:20.365 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=12
05:58:20.365 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=12 value=12
05:58:20.365 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=23
05:58:20.365 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=23 value=24
05:58:20.403 -> Debug:   es8388_set_bits_per_sample
05:58:20.403 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=12
05:58:20.403 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=12 value=12
05:58:20.403 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=23
05:58:20.403 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=23 value=24
05:58:20.403 -> Debug:   es8388_set_voice_volume: 20
05:58:20.403 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=46 value=6
05:58:20.441 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=47 value=6
05:58:20.441 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=48 value=0
05:58:20.441 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=49 value=0
05:58:20.441 -> Info:    i2s_set_pin
05:58:20.441 -> Debug:   audio_hal_ctrl_codec
05:58:20.441 -> Info:    Codec mode is 2, Ctrl:1
05:58:20.441 -> Debug:   es8388_ctrl_state
05:58:20.441 -> Debug:   es8388_start
05:58:20.441 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=43
05:58:20.478 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=43 value=128
05:58:20.478 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=43
05:58:20.478 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=4 value=60
05:58:20.478 -> Debug:   es8388_set_voice_mute
05:58:20.478 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=25
05:58:20.478 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=25 value=0
05:58:20.478 -> Debug:   es8388_start default is mode:2
05:58:20.512 -> Debug:   start default is decode mode:2
05:58:20.512 -> [I] AudioKit.h : 196 - Channels 2->2
05:58:20.512 -> [I] AudioTypes.h : 48 - sample_rate: 44100
05:58:20.512 -> [I] AudioTypes.h : 49 - channels: 2
05:58:20.512 -> [I] AudioTypes.h : 50 - bits_per_sample: 16
05:58:20.512 -> [I] AudioStreamsConverter.h : 336 - no bit combination: 16 -> 16
05:58:20.512 -> [I] AudioKit.h : 585 - void audio_tools::AudioKitStream::setupActions()
05:58:20.512 -> [W] AudioKit.h : 597 - Mode Button ignored because of conflict: 13 
05:58:20.550 -> [I] AudioActions.h : 42 - ActionLogic::add pin: 39 / logic: 2
05:58:20.550 -> [I] AudioActions.h : 55 - pin 39 -> INPUT
05:58:20.550 -> [I] AudioKit.h : 337 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
05:58:20.550 -> [I] AudioActions.h : 42 - ActionLogic::add pin: 18 / logic: 0
05:58:20.550 -> [I] AudioActions.h : 52 - pin 18 -> INPUT_PULLUP
05:58:20.550 -> [I] AudioKit.h : 337 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
05:58:20.583 -> [I] AudioActions.h : 42 - ActionLogic::add pin: 5 / logic: 0
05:58:20.583 -> [I] AudioActions.h : 52 - pin 5 -> INPUT_PULLUP
05:58:20.583 -> Debug:   audio_hal_set_volume: 100
05:58:20.583 -> Debug:   es8388_set_voice_volume: 100
05:58:20.583 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=46 value=33
05:58:20.583 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=47 value=33
05:58:20.620 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=48 value=0
05:58:20.620 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=49 value=0
05:58:20.620 -> [I] URLStream.h : 78 - virtual bool audio_tools::URLStreamDefault::begin(const char*, const char*, MethodID, const char*, const char*): http://stream.srg-ssr.ch/m/rsj/mp3_128
05:58:20.620 -> [I] Url.h : 73 - Url::parse
05:58:20.620 -> [I] Url.h : 109 - url->http://stream.srg-ssr.ch/m/rsj/mp3_128
05:58:20.654 -> [I] Url.h : 110 - host->stream.srg-ssr.ch
05:58:20.654 -> [I] Url.h : 111 - protocol->http
05:58:20.654 -> [I] Url.h : 112 - path->/m/rsj/mp3_128
05:58:20.654 -> [I] Url.h : 113 - port->80
05:58:20.756 -> .
05:58:21.749 -> [I] URLStream.h : 253 - WiFiClient
05:58:21.749 -> [I] HttpRequest.h : 193 - process connecting to host stream.srg-ssr.ch port 80
05:58:21.820 -> [I] HttpRequest.h : 181 - connected 1 timeout 60000
05:58:21.855 -> [I] HttpRequest.h : 204 - Free heap: 195416
05:58:21.855 -> [I] HttpHeader.h : 357 - HttpRequestHeader::setValues - path: /m/rsj/mp3_128
05:58:21.855 -> [I] HttpHeader.h : 254 - HttpHeader::write
05:58:21.855 -> [I] HttpHeader.h : 366 - HttpRequestHeader::write1stLine
05:58:21.855 -> [I] HttpHeader.h : 381 - -> GET /m/rsj/mp3_128 HTTP/1.1
05:58:21.855 -> [I] HttpHeader.h : 191 -  -> Host: stream.srg-ssr.ch 
05:58:21.855 -> [I] HttpHeader.h : 191 -  -> Connection: close 
05:58:21.855 -> [I] HttpHeader.h : 191 -  -> Accept: audio/mp3 
05:58:21.888 -> [I] HttpHeader.h : 297 -  -> <CR LF> 
05:58:21.888 -> [I] HttpRequest.h : 227 - Request written ... waiting for reply
05:58:21.888 -> [I] HttpHeader.h : 224 - HttpHeader::read
05:58:21.888 -> [W] HttpHeader.h : 231 - Waiting for data...
05:58:22.376 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> HTTP/1.1 200 OK
05:58:22.414 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Content-Type: audio/mpeg
05:58:22.414 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Date: Sun, 26 Jun 2022 03:58:21 GMT
05:58:22.414 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> icy-br:128
05:58:22.414 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> ice-audio-info: bitrate=128
05:58:22.414 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> icy-br:128
05:58:22.414 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> icy-description:SwissJazz_MP3@128Kbps
05:58:22.447 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> icy-genre:Entertainment
05:58:22.447 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> icy-mimetype:audio/mpeg
05:58:22.447 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> icy-name:/SwissJazz_Room2
05:58:22.447 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> icy-pub:0
05:58:22.447 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> icy-url:www.srgssr.ch
05:58:22.481 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Server: Icecast 2.4.0-kh10
05:58:22.481 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Cache-Control: no-cache, no-store
05:58:22.481 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Access-Control-Allow-Origin: *
05:58:22.481 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Access-Control-Allow-Headers: Origin, Accept, X-Requested-With, Content-Type
05:58:22.515 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Access-Control-Allow-Methods: GET, OPTIONS, HEAD
05:58:22.515 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Connection: Close
05:58:22.515 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> Expires: Mon, 26 Jul 1997 05:00:00 GMT
05:58:22.515 -> [I] HttpHeader.h : 161 - HttpHeader::readLine -> 
05:58:22.515 -> [W] HttpRequest.h : 153 - no CONTENT_LENGTH found in reply
05:58:22.515 -> [I] URLStream.h : 93 - size: 0
05:58:22.548 -> [I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
05:58:22.548 -> [I] AudioKit.h : 252 - virtual void audio_tools::AudioKitStream::setAudioInfo(audio_tools::AudioBaseInfo)
05:58:22.548 -> [I] I2SConfig.h : 95 - rx/tx mode: TX
05:58:22.548 -> [I] I2SConfig.h : 96 - port_no: 0
05:58:22.584 -> [I] I2SConfig.h : 97 - is_master: Master
05:58:22.584 -> [I] I2SConfig.h : 98 - sample rate: 48000
05:58:22.584 -> [I] I2SConfig.h : 99 - bits per sample: 16
05:58:22.584 -> [I] I2SConfig.h : 100 - number of channels: 2
05:58:22.584 -> [I] I2SConfig.h : 101 - i2s_format: I2S_STD_FORMAT
05:58:22.584 -> [I] I2SConfig.h : 103 - auto_clear:0
05:58:22.584 -> [I] I2SConfig.h : 110 - buffer_count:10
05:58:22.584 -> [I] I2SConfig.h : 111 - buffer_size:512
05:58:22.584 -> [I] I2SConfig.h : 115 - pin_bck: 14
05:58:22.584 -> [I] I2SConfig.h : 116 - pin_ws: 15
05:58:22.617 -> [I] I2SConfig.h : 117 - pin_data: -1515870811
05:58:22.617 -> [I] AudioTypes.h : 48 - sample_rate: 48000
05:58:22.617 -> [I] AudioTypes.h : 49 - channels: 2
05:58:22.617 -> [I] AudioTypes.h : 50 - bits_per_sample: 16
05:58:22.617 -> Debug:   audio_hal_codec_iface_config
05:58:22.617 -> Debug:   es8388_config_i2s
05:58:22.617 -> Debug:   es8388_config_fmt
05:58:22.617 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=12
05:58:22.617 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=12 value=12
05:58:22.651 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=23
05:58:22.651 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=23 value=24
05:58:22.651 -> Debug:   es8388_set_bits_per_sample
05:58:22.651 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=12
05:58:22.651 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=12 value=12
05:58:22.651 -> Debug:   i2c_bus_read_bytes: addr=32 reglen=1 datalen=1 - reg=23
05:58:22.687 -> Debug:   i2c_bus_write_bytes: addr=32 reglen=1 datalen=1 - reg=23 value=24
05:58:22.687 -> Debug:   write: 4608
05:58:22.721 -> Debug:   write: 4608
```

**Comment by billy8119** [comment]
I was getting 6 also.  I notice that later on, you set volume to "100" aka 33.  That looks right, IE bug free.  I may be going crazy, but I thought I tried the experiment setting volume to 100, and it was still silent as a mouse.  I haven't tried since I put the debug options on, since I was under the impression that setVolume wanted a float from 0 to 1.. which is clearly not the case (default in the original code is set to 0.7) Just so you don't think I'm a TOTAL idiot, I got this idea from your earlier post suggesting "player.setVolume(1.0)" would be max volume.
Well, at least I know I have some volume control options, I just need to sort out what to set it to.  I also set DACregister 4 and 5 to "0" to set attenuation at the lowest possible levels.  I haven't measured if it makes a difference yet.
Am I correct in assuming that kit.setVolume(100) sets the kit opamps on? I think this is automatically disabled when the system detects headphones are plugged in, yeah? With "selfAudioKit->setSpeakerActive(powerActive);" from AudioKitHAL.h...?  I might try playing with that to see if those opamps are useful when enabled, regardless of headphone insertion.

**Comment by pschatzmann** [comment]
The PA is activated on init. There is an automatic headphone detection if you call processActions() in the loop, which is deactivating the speakers when the headphones are inserted.

If you want both: that's the default logic and you don't need to do anything...

**Reply by billy8119** [reply]
Oh.  Ok. Much appreciated.  So now, with kit,setVolume(100) and also player.setVolume(100) it's very loud, and the signal is mostly intact (there's a little distortion, which is to be expected) THANK-YOU!! 
Also, in addition to thanking you for all the help, thank-you very much for the jaw-droppingly fabulous code; you shaved at least a month of me fumbling around off of my project! (I'm making a sound-enhanced robot costume for my son :)
So...EPIC THANK-YOUs!

**Comment by pschatzmann** [comment]
Cool: with my latest commit from today you can also use kit.setVolume(1.0) to set the max volume (as an alternative to kit.setVolume(100)).

I am not sure what a sound-enhanced robot costume would exactly do, but maybe the [arduino-simple-tts](https://github.com/pschatzmann/arduino-simple-tts) might be if interest to play some prerecorded sounds. Or my other TTS libraries to generate some robot voice...

**Comment by billy8119** [comment]
Out of curiosity, in your travels, did you ever find a proper schematic? The schematic offered doesn't have the interconnection between esp32 and es8388, it just has both the esp32 AND the esp32-module-with-es8388, but no specific details on how the es8388 interconnects with the esp32 and other peripherals. For example, what's connected to DAC2? I dunno!

**Comment by pschatzmann** [comment]
So far I found the following:
- https://docs.ai-thinker.com/_media/esp32-a1s_v2.3_specification.pdf
- https://docs.ai-thinker.com/_media/esp32-audio-kit_v2.2_sch.pdf

The problem is that there are at least 6 different variants with different pin assignments and I never found any proper documentation that would relate to the variant...
